File: LinkedList Java and Kotlin/Middle of the Linked List/IterativeSolution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        // Use head as slow pointer
        // Use fast as fast pointer
        ListNode fast=head;
        while(fast!=null && fast.next!=null) {
            head=head.next;
            fast=fast.next.next;    
        }
        
        return head;
    }
}

File: LinkedList Java and Kotlin/Middle of the Linked List/IterativeSolution.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun middleNode(head: ListNode?): ListNode? {
        //slow and fast pointers
        var slow:ListNode? = head
        var fast:ListNode? = head
        
        while(fast!=null && fast?.next!=null) {
            slow = slow!!.next
            fast = fast.next.next
        }
        return slow
    }
}
